File: app/src/androidMain/kotlin/com/cs407/cubemaster/ui/screens/PermissionScreen.kt
package com.cs407.cubemaster.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.navigation.NavController
import androidx.navigation.compose.rememberNavController
import com.cs407.cubemaster.ui.theme.CubemasterTheme

@Composable
fun PermissionScreen(
    modifier: Modifier = Modifier,
    navController: NavController,
    permissionGranted: Boolean,
    showPermissionDeniedMessage: Boolean,
    requestPermission: () -> Unit,
    openSettings: () -> Unit
) {
    Box(modifier = modifier.fillMaxSize()) {
        Column(
            modifier = modifier
                .fillMaxSize()
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Text(
                text = "Camera Permission Required",
                style = MaterialTheme.typography.headlineMedium,
                color = Color.White,
                textAlign = TextAlign.Center
            )
            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = when {
                    permissionGranted -> "Permission Granted!"
                    showPermissionDeniedMessage -> "Camera permission is required to scan the cube. Please enable it in the app settings."
                    else -> "This app needs camera access to scan your Rubik's Cube. Please grant the permission to continue."
                },
                style = MaterialTheme.typography.bodyLarge,
                color = Color.White,
                textAlign = TextAlign.Center
            )
            Spacer(modifier = Modifier.height(32.dp))
            if (showPermissionDeniedMessage) {
                Button(onClick = openSettings) {
                    Text(text = "Open Settings")
                }
            } else {
                Button(onClick = requestPermission, enabled = !permissionGranted) {
                    Text(text = "Grant Permission")
                }
            }
        }
        Button(
            onClick = { navController.popBackStack() },
            modifier = Modifier.align(Alignment.BottomStart).padding(16.dp)
        ) {
            Text(text = "Back")
        }
        if (permissionGranted) {
            Button(
                onClick = { navController.navigate("scan") },
                modifier = Modifier.align(Alignment.BottomEnd).padding(16.dp)
            ) {
                Text(text = "Next")
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
fun PermissionScreenPreview() {
    CubemasterTheme {
        PermissionScreen(
            navController = rememberNavController(),
            permissionGranted = false,
            showPermissionDeniedMessage = false,
            requestPermission = {},
            openSettings = {}
        )
    }
}

@Preview(showBackground = true)
@Composable
fun PermissionScreenGrantedPreview() {
    CubemasterTheme {
        PermissionScreen(
            navController = rememberNavController(),
            permissionGranted = true,
            showPermissionDeniedMessage = false,
            requestPermission = {},
            openSettings = {}
        )
    }
}

@Preview(showBackground = true)
@Composable
fun PermissionScreenDeniedPreview() {
    CubemasterTheme {
        PermissionScreen(
            navController = rememberNavController(),
            permissionGranted = false,
            showPermissionDeniedMessage = true,
            requestPermission = {},
            openSettings = {}
        )
    }
}

File: iOS/Cubemaster/Views/PermissionScreen.swift
import SwiftUI
import AVFoundation

class PermissionManager: ObservableObject {
    @Published var permissionGranted: Bool = false
    @Published var showPermissionDeniedMessage: Bool = false
    
    init() {
        checkPermissionStatus()
    }
    
    func checkPermissionStatus() {
        let status = AVCaptureDevice.authorizationStatus(for: .video)
        permissionGranted = status == .authorized
        showPermissionDeniedMessage = status == .denied
    }
    
    func requestPermission() {
        AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
            DispatchQueue.main.async {
                self?.permissionGranted = granted
                if !granted {
                    // Check if we should show the denied message
                    // On iOS, if status is denied, user must go to settings
                    let status = AVCaptureDevice.authorizationStatus(for: .video)
                    self?.showPermissionDeniedMessage = status == .denied
                }
            }
        }
    }
    
    func openSettings() {
        if let settingsUrl = URL(string: UIApplication.openSettingsURLString) {
            UIApplication.shared.open(settingsUrl)
        }
    }
}

struct PermissionScreen: View {
    @EnvironmentObject var permissionManager: PermissionManager
    @Environment(\.dismiss) var dismiss
    @State private var navigationPath = NavigationPath()
    
    var body: some View {
        NavigationStack(path: $navigationPath) {
            ZStack {
                // Gradient background matching Android
                LinearGradient(
                    colors: [Colors.lightOrange, Colors.darkOrange],
                    startPoint: .top,
                    endPoint: .bottom
                )
                .ignoresSafeArea()
                
                // Main content column (16dp padding, centered)
                VStack(spacing: 0) {
                    Spacer()
                    
                    Text("Camera Permission Required")
                        .font(.title2) // headlineMedium equivalent
                        .fontWeight(.semibold)
                        .foregroundColor(.white)
                        .multilineTextAlignment(.center)
                    
                    Spacer()
                        .frame(height: 16)
                    
                    Text(permissionText)
                        .font(.body) // bodyLarge equivalent
                        .foregroundColor(.white)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 16)
                    
                    Spacer()
                        .frame(height: 32)
                    
                    // Conditional button based on permission state
                    if permissionManager.showPermissionDeniedMessage {
                        Button(action: {
                            permissionManager.openSettings()
                        }) {
                            Text("Open Settings")
                                .font(.headline)
                                .foregroundColor(.white)
                                .frame(maxWidth: .infinity)
                                .padding()
                                .background(Color.blue)
                                .cornerRadius(12)
                        }
                        .padding(.horizontal, 16)
                    } else {
                        Button(action: {
                            permissionManager.requestPermission()
                        }) {
                            Text("Grant Permission")
                                .font(.headline)
                                .foregroundColor(.white)
                                .frame(maxWidth: .infinity)
                                .padding()
                                .background(permissionManager.permissionGranted ? Color.gray : Color.blue)
                                .cornerRadius(12)
                        }
                        .disabled(permissionManager.permissionGranted)
                        .padding(.horizontal, 16)
                    }
                    
                    Spacer()
                }
                .padding(16)
                
                // Bottom buttons overlay
                VStack {
                    Spacer()
                    HStack {
                        // Back button (bottom-left, 16dp padding)
                        Button(action: {
                            dismiss()
                        }) {
                            Text("Back")
                                .font(.headline)
                                .foregroundColor(.white)
                                .padding()
                                .background(Color.blue)
                                .cornerRadius(12)
                        }
                        .padding(16)
                        
                        Spacer()
                        
                        // Next button (bottom-right, 16dp padding, only when permission granted)
                        if permissionManager.permissionGranted {
                            Button(action: {
                                navigationPath.append("scan")
                            }) {
                                Text("Next")
                                    .font(.headline)
                                    .foregroundColor(.white)
                                    .padding()
                                    .background(Color.blue)
                                    .cornerRadius(12)
                            }
                            .padding(16)
                        }
                    }
                }
            }
            .navigationDestination(for: String.self) { destination in
                if destination == "scan" {
                    ScanScreen()
                }
            }
        }
    }
    
    private var permissionText: String {
        if permissionManager.permissionGranted {
            return "Permission Granted!"
        } else if permissionManager.showPermissionDeniedMessage {
            return "Camera permission is required to scan the cube. Please enable it in the app settings."
        } else {
            return "This app needs camera access to scan your Rubik's Cube. Please grant the permission to continue."
        }
    }
}

